tests/qtrajectory_test.cc

```diff
@@ -0,0 +1,439 @@
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.measure(*[qs[0], qs[1]], key='m'),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+)
+
+reps = 10000000
+
+sim = cirq.Simulator()
+res = sim.run(ncircuit, repetitions=reps)
+
+for key, val in sorted(res.histogram(key='m').items()):
+  print(f'{key} {float(val) / reps}')
+
+*/
+
+  std::vector<double> expected_results = {
+    0.412300, 0.230500, 0.057219, 0.299982,
+  };
+
+  auto ncircuit = GenerateNoisyCirquit(0.02, AddPhaseDumpNoise);
+  Run2(ncircuit, expected_results);
+}
+
+TEST(QTrajectoryTest, AmplDump) {
+/* The expected results are obtained with the following Cirq code.
+
+import cirq
+
+qs = cirq.LineQubit.range(2)
+
+channel = cirq.amplitude_damp(0.05)
+
+ncircuit = cirq.Circuit(
+  cirq.H(qs[0]),
+  cirq.H(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.rx(0.7)(qs[0]),
+  cirq.ry(0.1)(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ry(0.4)(qs[0]),
+  cirq.rx(0.7)(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.measure(*[qs[0], qs[1]], key='m'),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+)
+
+reps = 10000000
+
+sim = cirq.Simulator()
+res = sim.run(ncircuit, repetitions=reps)
+
+for key, val in sorted(res.histogram(key='m').items()):
+  print(f'{key} {float(val) / reps}')
+
+*/
+
+  std::vector<double> expected_results = {
+    0.500494, 0.235273, 0.090879, 0.173354,
+  };
+
+  auto ncircuit = GenerateNoisyCirquit(0.05, AddAmplDumpNoise);
+  Run1(ncircuit, expected_results);
+}
+
+TEST(QTrajectoryTest, GenDump) {
+/* The expected results are obtained with the following Cirq code.
+
+import cirq
+
+qs = cirq.LineQubit.range(2)
+
+channel = cirq.generalized_amplitude_damp(0.5, 0.1)
+
+ncircuit = cirq.Circuit(
+  cirq.H(qs[0]),
+  cirq.H(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.rx(0.7)(qs[0]),
+  cirq.ry(0.1)(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ry(0.4)(qs[0]),
+  cirq.rx(0.7)(qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.ISWAP(qs[0], qs[1]),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+  cirq.measure(*[qs[0], qs[1]], key='m'),
+  channel.on(qs[0]),
+  channel.on(qs[1]),
+)
+
+reps = 10000000
+
+sim = cirq.Simulator()
+res = sim.run(ncircuit, repetitions=reps)
+
+for key, val in sorted(res.histogram(key='m').items()):
+  print(f'{key} {float(val) / reps}')
+
+*/
+
+  std::vector<double> expected_results = {
+    0.318501, 0.260538, 0.164616, 0.256345,
+  };
+
+  auto ncircuit = GenerateNoisyCirquit(0.1, AddGenAmplDumpNoise);
+  Run2(ncircuit, expected_results);
+}
+
+
+}  // namespace qsim
```

Comment:
It would be nice to also have a test case verifying that the final state produced by the trajectory simulator is correct, or to update `Run1` and `Run2` to perform such a check.

Reply:
What do you mean by "correct"? The final state after each iteration is the result of some stochastic process and it is random. I guess one way to check for correctness is to run a clean circuit and to compare with the clean circuit simulator. I added such a test.

Reply:
You make a good point - I think the clean-circuit test covers the concerns I had in the original comment.